Fix avatar generator centering

skjnldsv opened the pull request:
Fix #8841 

@nextcloud/designers @biva 

| Before | After |
|:-------:|:------:|
| ![avatar 64 5](https://user-images.githubusercontent.com/14975046/38799419-c612edd6-4164-11e8-830b-70547d52d629.png) | ![avatar 64](https://user-images.githubusercontent.com/14975046/38799413-c546dd4a-4164-11e8-99dd-464a3e540b60.png) |
| ![avatar 64 4](https://user-images.githubusercontent.com/14975046/38799418-c5f70922-4164-11e8-877f-c9d49c4aad60.png) | ![avatar 64 1](https://user-images.githubusercontent.com/14975046/38799415-c588807e-4164-11e8-94f5-f07c2e14fca2.png) |
| ![avatar 64 3](https://user-images.githubusercontent.com/14975046/38799417-c5cbc6ae-4164-11e8-9cbb-0ee8a52c4603.png) | ![avatar 64 2](https://user-images.githubusercontent.com/14975046/38799416-c5a79aa4-4164-11e8-9a06-66ca49a737bc.png) |
| ![avatar 64 6](https://user-images.githubusercontent.com/14975046/38799420-c6334b08-4164-11e8-8b65-4f1c5ad93270.png) | ![avatar 64 7](https://user-images.githubusercontent.com/14975046/38799421-c651d58c-4164-11e8-893a-603bbb404a83.png) |

skjnldsv committed:
Fix avatar generator centering

Signed-off-by: John Molakvoæ (skjnldsv) <skjnldsv@protonmail.com>

MorrisJobke commented on lib/private/Avatar.php:
```diff
@@ -291,6 +288,35 @@ private function generateAvatar($userDisplayName, $size) {
+		$xr = abs(max($box[2], $box[4]));
+		$yr = abs(max($box[5], $box[7]));
+
+		// calculate bottom left plcaement
```
Typo ;)

skjnldsv commented:
> I don't see a difference here (1 and 3 were created on master, 2 and 4 were created on this branch)

Are you sure the cache was cleared and stuff, because if you look at my first post, you'll see we really have a difference :)
Nonetheless, the centering still ignores the accent, so the letters are still a bit too low, I'll check what I can do

danxuliu reviewed (commented):
I can see a difference between avatars generated in _master_ branch and avatars generated in this branch. In general they look better now :-) (although for some specific characters they were better before, like for _A_).

Unfortunately I do not think that it will be possible to get perfect centering with small sizes if sub-pixel accuracy is not provided by the underlying libraries :-( (@skjnldsv you cheated a little with the examples in the first post, as with 64x64px the avatar can be better centered than with 32x32px :-P ). Anyway, this looks good enough ;-)

skjnldsv commented:
> (@skjnldsv you cheated a little with the examples in the first post, as with 64x64px the avatar can be better centered than with 32x32px :-P ). Anyway, this looks good enough ;-)

Damn, busted! :D 
Let put this pr on hold, and let me see if I can find a way of properly generate them :)
Especially if I can respect the font baseline instead of box ;)

rullzer commented:
Maybe just generate the avatar at 128x128 and scale it down :P?

skjnldsv commented:
@rullzer actually, I think that's how I'm heading. 
Generate svg file and convert to png :)

danxuliu commented:
Scaling the text down will probably cause it to be too blurry :-( Even with a SVG file you would need to somehow align the character to the grid so it looks crisp once rasterized (font hinting), although maybe that is automatically done by the library, no idea :-)

skjnldsv commented:
Looks pretty good to me :D ![avatar](https://user-images.githubusercontent.com/14975046/38815114-77a706bc-4193-11e8-9b8e-d4bb58966a68.png)

danxuliu commented:
>  Looks pretty good to me :D

Indeed; I was expecting it to be much worse. Fine by me then ;-)

MorrisJobke commented:
CI is fixed on master.

skjnldsv commented:
As a reference, here are our solution and their issues
1. Convert svg to png in server:
   - we need Imagemagick to convert svg, gd does not supports it
   - we can use phantomjs to trick the conversion, but  not very convenient and EOL has been announced for phantomjs
2. Pre-generate all avatar
   - We can't generate all non-latin chars. It would be fine  for latin only, but we need to keep support for japanese, idian... 
3. Serve svg
   - Web supports it perfectly
   - Lighter than generating png
   - IOS & Android have some issues with it. Currently @Ivansss and @tobiasKaminsky are looking at our option this week and we'll see what we can do later
4. Serve png AND svg: If svg supported: serve svg if not, serve png
   - Very bad ux since we will have good avatar and 'bad' ones

tobiasKaminsky commented:
I just took a look into our code and we already using svg rendering on activity/notification list.
As the images are cached locally, I hope that there is no big speed drawback.

All in all I would like to give svg avatars a try. 
Currently we are not using avatars that often.

Ivansss commented:
iOS is not SVG-friendly (SVG images are not supported by default).
We could use third-party libraries to render SVGs, but the performance/user experience would be definitely affected.
So I would go with the 4th option or the 1st one if possible.
Since the server generate and serve the avatars, I think it should also convert them to png.

skjnldsv commented:
@Ivansss how does the avatar looks on ios right now? Are the letters shifted to the right or the left? Is it centered?

rullzer commented:
We won't do the 1st option. We disabled the required dependency on imagemagik for a reason ;)

Ivansss commented:
Sorry @skjnldsv , I totally missed your message.
They look like this at the moment.
![img_0148](https://user-images.githubusercontent.com/4638605/39300555-63eb57de-494c-11e8-9b09-48e8dc2532df.PNG)
So a bit shifted to the right? Depends on the character I would say.
Could the 1st option be done using another library instead of imagemagick?

skjnldsv commented:
@Ivansss no problem :)
Can you try with the "É" letter?

Ivansss commented:
![img_e0149](https://user-images.githubusercontent.com/4638605/39300894-8da117a2-494d-11e8-9f3f-ff9e7e16b387.JPG)

biva commented:
:(
É should be at the exact same place as E. We shouldn't take the accent into account

skjnldsv commented:
@biva that's the whole point of this pr ^^
@Ivansss thanks, we will introduce the fallback to png, at least the letter is better centered now, but the accent is still taken in consideration in the png generation :/

biva commented:
@skjnldsv sorry ! I thought it was the new version. I let you work guys, sorry again.

skjnldsv commented:
@biva no need to apologise :hugs: :)

This is the new version for the png generation. We will nonetheless introduce the svg avatar as well as primary avatar for supported systems (all except ios ^^) which have the letters all properly aligned ;)

see https://github.com/nextcloud/server/pull/9197#issuecomment-381619266

skjnldsv committed:
Return Svg avatars

Signed-off-by: John Molakvoæ (skjnldsv) <skjnldsv@protonmail.com>

skjnldsv committed:
Use base image size

Signed-off-by: John Molakvoæ (skjnldsv) <skjnldsv@protonmail.com>

skjnldsv committed:
Controller tests fixes

Signed-off-by: John Molakvoæ (skjnldsv) <skjnldsv@protonmail.com>

skjnldsv committed:
Unify colour algorithm output

Signed-off-by: John Molakvoæ (skjnldsv) <skjnldsv@protonmail.com>

skjnldsv committed:
Added new tests

Signed-off-by: John Molakvoæ (skjnldsv) <skjnldsv@protonmail.com>

skjnldsv committed:
Fixed since requirement

Signed-off-by: John Molakvoæ (skjnldsv) <skjnldsv@protonmail.com>

MorrisJobke commented:
@skjnldsv Is this ready for review?

skjnldsv commented:
@MorrisJobke Yes, I was waiting for the tests :)
Let's go! :writing_hand:

MorrisJobke commented:
I don't have "Open Sans" installed on my system and then the avatar looks like this:

![bildschirmfoto 2018-05-07 um 16 27 16](https://user-images.githubusercontent.com/245432/39707070-91ef2c62-5213-11e8-9cbe-b3b96d8c2174.png)

skjnldsv commented:
@MorrisJobke damn it, that's something we did not think of :/

MorrisJobke commented:
> @MorrisJobke damn it, that's something we did not think of :/

Render it on the server side and then generate a PNG out of it? Then also the iOS app will work with this avatar without problems.

pixelipo commented:
Those are rendered as SVG? CSS would solve both the font issue and the bottom alignment. What am I missing here?

skjnldsv commented:
@pixelipo ! Welcome back :hugs: :)
Yes, that's why we're switching to svg avatars, because the css allow us to properly align the letters from the baseline.
The only issue is the font. A svg loaded as img, does not take the page css font-face in account :/

skjnldsv commented:
Because it's being served as a image/xml+svg for our other devices. And it's easier to just use the userid in url `http://xxx/avatar/userid/size` in an `img` tag than use javascript to load data and stuff, or implement if on our page directly.

This way we only have one get point for th avatar and only one location to edit if we want to change the process (like we're doing right now) instead of everywhere where the div would have been use :)

skjnldsv committed:
Embedded font

Signed-off-by: John Molakvoæ (skjnldsv) <skjnldsv@protonmail.com>

skjnldsv commented:
@MorrisJobke can you try again? :)

MorrisJobke commented:
Still doesn't show up properly:

<img width="389" alt="bildschirmfoto 2018-05-23 um 14 02 53" src="https://user-images.githubusercontent.com/245432/40423199-0aaa8e64-5e92-11e8-82f3-42a2bb7319ce.png">

skjnldsv committed:
Avatar imagick bump

Signed-off-by: John Molakvoæ (skjnldsv) <skjnldsv@protonmail.com>

skjnldsv committed:
Removed old route, fix tests and fix var type

Signed-off-by: John Molakvoæ (skjnldsv) <skjnldsv@protonmail.com>

skjnldsv commented:
@nextcloud/designers Okay
I changed our way of doing. If imagick is installed, the svg avatar will be used and converted to png.
If not, the old (but improved) png gd avatar will be used. centered horizontally but can have some glitches with accents or custom chars.

It's better than nothing though :)

MorrisJobke commented:
Works now, but the initial goal of centering letters with special signs above or below them is still not reached:

<img width="351" alt="bildschirmfoto 2018-05-28 um 15 44 17" src="https://user-images.githubusercontent.com/245432/40617304-4085ce9a-628e-11e8-96a3-ecd705e0a601.png">

I'm fine with getting this in due to the color re-calculation nevertheless.

skjnldsv commented:
@MorrisJobke 

> I changed our way of doing. If imagick is installed, the svg avatar will be used and converted to png.
If not, the old (but improved) png gd avatar will be used. centered horizontally but can have some glitches with accents or custom chars.

You probably don't have imagick set up :)

MorrisJobke commented:
> You probably don't have imagick set up :)

It says that it is installed:

![bildschirmfoto 2018-05-28 um 15 53 59](https://user-images.githubusercontent.com/245432/40617687-627d12fa-628f-11e8-9543-3ea88cb88928.png)

skjnldsv commented:
@MorrisJobke maybe a cache issue?
Try getting a new avatar size?
`https://yourdomain.com/avatar/Test/100`

skjnldsv commented:
Any news?

skjnldsv committed:
Fix indent

Signed-off-by: John Molakvoæ (skjnldsv) <skjnldsv@protonmail.com>

MorrisJobke commented:
Restarted CI - then this can go in

skjnldsv committed:
Removed debug stack

Signed-off-by: John Molakvoæ (skjnldsv) <skjnldsv@protonmail.com>

skjnldsv commented:
@MorrisJobke Here we go! :see_no_evil:
